storage/innobase/include/trx0trx.h

```diff
@@ -1178,10 +1194,19 @@ struct trx_t : ilist_node<>
 
   /** Do the bulk insert for the buffered insert operation
   for the transaction.
+  @param ddl Apply buffered insert only when bulk_insert is
+             TRX_DDL_BULK
   @return DB_SUCCESS or error code */
-  dberr_t bulk_insert_apply()
+  dberr_t bulk_insert_apply(bool ddl=false)
   {
-    return UNIV_UNLIKELY(bulk_insert) ? bulk_insert_apply_low(): DB_SUCCESS;
+    if (ddl)
+      return bulk_insert == TRX_DDL_BULK
+        ? bulk_insert_apply_low()
+        : DB_SUCCESS;
+
+    return bulk_insert == TRX_DML_BULK
+      ? bulk_insert_apply_low()
+      : DB_SUCCESS;
   }
```

Comment:
It seems possible that the function call may be duplicated in the expansion of this inline function. Can we somehow simplify this, to allow the function parameter `ddl` to be compared to the data member `bulk_insert` directly?
```c++
  dberr_t bulk_insert_apply(trx_bulk_insert type)
  {
    ut_ad(type != TRX_NO_BULK);
    return bulk_insert == type ? bulk_insert_apply_low() : DB_SUCCESS;
  }
```
We might even make this a template parameter, to have a compile-time assertion:
```c++
  template<trx_bulk_insert type=TRX_DML_BULK>
  dberr_t bulk_insert_apply()
  {
    static_assert(type != TRX_NO_BULK, "");
    return bulk_insert == type ? bulk_insert_apply_low() : DB_SUCCESS;
  }
```